Make options sufficiently similar to the Python plugin

rck opened the pull request:
This implements #4 :
- introduce proper size handling
- make options similar/compatible enough

Please see the individual commit messages for more details.

rck committed:
handle sizes with units

This adds units to the Size option and does the usual checks. If not set
at all, set to 100MB (default from the Python based plugin). I did not
care to make MB the default if no unit is specified. This is a trade-off
I am willing to take. The plugin should be "sufficiently similar" to the
Python one, but it does not need to be a pixel perfect copy.

This implements one part of #4.

rck committed:
let LINSTOR do mkfs, add FSOpts

The Python based plugin has "fsopts", where one can specify additional
parameters. The problem is that we did not have that part under control
by using the k8s mounter utils. The good thing is that LINSTOR itself
can create a file system on top of the device.

- use LINSTOR to create the FS. This is a good idea in general, LINSTOR
only supports xfs and ext4, but it already sets some mkfs options
accordingly.
- add FSOpts for additional options.
- use only the Mount()er part from k8s.

Again, this does not make the plugin pixel perfect to the Python one,
but IMO with the previous commit, we can now:

Close #4

beornf commented on driver.go:
```diff
@@ -156,8 +159,32 @@ func (l *LinstorDriver) newParams(name string, options map[string]string) (*Lins
+	if params.FS == "" {
+		params.FS = "ext4"
+	}
+
```
Moving my replica options check `params.Replicas == 0 { ... }` might be cleaner here see lines 222-224

rck replied:
mhm. will add that on top I guess. I will send an updated PR soon

beornf commented on driver.go:
```diff
@@ -294,7 +325,7 @@ func (l *LinstorDriver) Mount(req *volume.MountRequest) (*volume.MountResponse,
 		return nil, err
 	}
-	err = l.mounter.FormatAndMount(source, target, params.FS, params.MountOpts)
+	err = l.mounter.Mount(source, target, params.FS, params.MountOpts)
```
We should reference the `FileSystem/Type` property here as the mount call receives global config options only.

rck replied:
good point. why did that work? :D

beornf replied:
I saw in linstor-csi the resource options were serialized into an Aux property so it could be read in mount but I wasn't really sure what format to use..

rck committed:
mv Replicas check to newParams()

newParams() is now used to set up the configuration including defaults.
Therfore move this one also to newParams.

beornf commented:
Merging the commits for the last open issue. Time to do some testing?

rck commented:
Testing is always good, right? :D. There are some minor things I'd like to change a bit, like the `Makefile`, maybe a PR where I add the README/man-page generation from the Python based one,... but for the core functionality I think we are close. If I don't find anything new, then I will let some of my colleagues test it next week and we progress from there.

beornf commented:
I've just tested the latest plugin and see that FileSystem/Type is set to xfs in the resource definition. However the plugin is then unable to mount the drbd device as xfs:
```
Mounting command: mount
Mounting arguments: -t xfs /dev/drbd1000 /var/lib/docker-volumes/linstor/foobar
Output: mount: /var/lib/docker-volumes/linstor/foobar: wrong fs type, bad option, bad superblock on /dev/drbd1000, missing codepage or helper program, or other error.
```

Could this be a concurrency or property issue in Linstor? I'm testing with the latest v1.1.2 versions of Linstor satellite and controller.

beornf commented:
Actually seems this is not a Linstor issue as the plugin is working fine on version 9.0.19 of the DRBD kernel driver. I guess there must be some compatibility issue with version 9.0.20 of the DRBD kernel driver on my system.

rck commented:
That is weird. I tried that manually (`linstor` client), by setting the according property (`rd sp resname FileSystem/Type xfs`) and it mounted on the storage node and on a diskless node. Can you reproduce it with a cli sequence? I used 9.0.20 as released/tagged.

beornf commented:
I've just been testing on a new Linstor cluster with /proc/drbd:
```
version: 9.0.20-1 (api:2/proto:86-115)
GIT-hash: 7dce3c8be99f4912f1490f9bb37f5aff6c873335 build by ...
Transports (api:16): tcp (9.0.20-1)
```
The manual sequence of test commands are:
```
linstor rd create foobar
linstor rd set-property foobar FileSystem/Type xfs
linstor vd create foobar 100M
linstor r create node1 node2 foobar
mkdir /mnt/foobar
mount /dev/drbd1000 /mnt/foobar/
```
After attempting to mount the resource on node2 I got the error:
```
mount: wrong fs type, bad option, bad superblock on /dev/drbd1000,
       missing codepage or helper program, or other error
```
Also my Linux kernel version is `4.9.0-11-amd64` if this is any help.

rck commented:
kernels `4.[89].*` unfortunately are broken due to a bug in the DRBD compat layer. There is public branch that should fix this (https://github.com/LINBIT/drbd/tree/drbd-9.0-debian9), and there will be a DRBD 9.0.21-rc1 very soon

beornf commented:
Okay that makes sense. I'll hold off updating until the next release which should resolve the issue on Debian 9.
